File: action_challenge_py/action_challenge_py/move_one_axis_server.py
#!/usr/bin/env python3
import time
import threading
from math import ceil

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from rclpy.action.server import ServerGoalHandle
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup
from my_robot_interfaces.action import MoveOneAxis
 
 
class MoveOneAxisServerNode(Node):
    def __init__(self):
        super().__init__("move_one_axis_server")
        self.current_position_ = 50
        self.goal_handle_ : ServerGoalHandle = None
        self.goal_lock_ = threading.Lock()
        self.move_server_ = ActionServer(
            self,
            MoveOneAxis,
            "move_one_axis",
            goal_callback=self.goal_callback,
            handle_accepted_callback=self.handle_accepted_callback,
            cancel_callback=self.cancel_callback,
            execute_callback=self.execute_callback,
            callback_group=ReentrantCallbackGroup()
        )

        self.get_logger().info("Move One Axis Server has been started.")

    def goal_callback(self, goal_request: MoveOneAxis.Goal):
        self.get_logger().info("Goal request received")

        # Check the goal request
        if goal_request.target_position < 0 or goal_request.target_position > 100:
            self.get_logger().info("Goal request rejected. Target position must be between 0 and 100")
            return GoalResponse.REJECT
        if goal_request.velocity < 0:
            self.get_logger().info("Goal request rejected. Velocity must be positive")
            return GoalResponse.REJECT

        # Policy: preempt existing goal when receiving a new goal
        with self.goal_lock_:
            if self.goal_handle_ is not None and self.goal_handle_.is_active:
                self.get_logger().info("Abort current goal and accept new goal")
                self.goal_handle_.abort()

        self.get_logger().info("Goal request accepted")
        return GoalResponse.ACCEPT

    def handle_accepted_callback(self, goal_handle: ServerGoalHandle):
        goal_handle.execute()

    def cancel_callback(self, goal_handle: ServerGoalHandle):
        self.get_logger().info("Goal request cancelled")
        return CancelResponse.ACCEPT # or REJECT

    def execute_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            self.goal_handle_ = goal_handle

        feedback = MoveOneAxis.Feedback()
        result = MoveOneAxis.Result()

        # Get request from goal
        target_pos = goal_handle.request.target_position
        target_vel = goal_handle.request.velocity

        # Execute the action
        self.get_logger().info("Executing the goal")
        num_move = ceil(abs(target_pos-self.current_position_)/target_vel)
        for i in range(num_move):
            if not goal_handle.is_active:
                result.final_position = self.current_position_
                result.message = "Goal has been preempted"
                return result
            if goal_handle.is_cancel_requested:
                self.get_logger().info("Canceling the goal")
                goal_handle.canceled()
                result.final_position = self.current_position_
                result.message = "Goal has been canceled"
                return result

            if abs(target_pos-self.current_position_) < target_vel:
                self.current_position_ += target_pos-self.current_position_
            else:
                if target_pos > self.current_position_:
                    self.current_position_ += target_vel
                else:
                    self.current_position_ -= target_vel
            feedback.current_position = self.current_position_
            goal_handle.publish_feedback(feedback)
            self.get_logger().info("Current position: %d" %self.current_position_)
            time.sleep(1)

        # Once done, set goal final state
        goal_handle.succeed()

        # and send the result
        result.final_position = self.current_position_
        result.message = "Goal has been reached"
        return result
 
 
def main(args=None):
    rclpy.init(args=args)
    node = MoveOneAxisServerNode()
    rclpy.spin(node, MultiThreadedExecutor())
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: action_challenge_cpp/src/move_one_axis_server.cpp
#include <cmath>
#include <cstdlib>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/move_one_axis.hpp"

using MoveOneAxis = my_robot_interfaces::action::MoveOneAxis;
using MoveOneAxisGoalHandle = rclcpp_action::ServerGoalHandle<MoveOneAxis>;
using namespace std::placeholders;


class MoveOneAxisServerNode : public rclcpp::Node
{
public:
    MoveOneAxisServerNode() : Node("move_one_axis")
    {
        cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
        move_one_axis_server_ = rclcpp_action::create_server<MoveOneAxis>(
            this,
            "move_one_axis",
            std::bind(&MoveOneAxisServerNode::goal_callback, this, _1, _2),
            std::bind(&MoveOneAxisServerNode::cancel_callback, this, _1),
            std::bind(&MoveOneAxisServerNode::handle_accepted_callback, this, _1),
            rcl_action_server_get_default_options(),
            cb_group_
        );
        
        RCLCPP_INFO(this->get_logger(), "Move one axis server has been started");
    }
 
private:
    rclcpp_action::Server<MoveOneAxis>::SharedPtr move_one_axis_server_;
    rclcpp::CallbackGroup::SharedPtr cb_group_;
    rclcpp_action::GoalUUID preempted_goal_id_;
    std::mutex mutex_;
    std::shared_ptr<MoveOneAxisGoalHandle> goal_handle_;
    int current_position_{50};

    rclcpp_action::GoalResponse goal_callback(const rclcpp_action::GoalUUID &uuid, 
        std::shared_ptr<const MoveOneAxis::Goal> goal) {
        (void)uuid;
        RCLCPP_INFO(this->get_logger(), "Goal has been received");

        // Validate goal
        if ((goal->target_position < 0) || (goal->target_position > 100)) {
            RCLCPP_ERROR(this->get_logger(), "Goal has been rejected");
            return rclcpp_action::GoalResponse::REJECT;
        }
        if (goal->velocity < 0) {
            RCLCPP_ERROR(this->get_logger(), "Goal has been rejected");
            return rclcpp_action::GoalResponse::REJECT;
        }
        
        // Policy: preempt existing goal when receiving a new valid goal
        {
            std::lock_guard<std::mutex> lock(mutex_);
            if (goal_handle_) {
                if (goal_handle_->is_active()) {
                    RCLCPP_WARN(this->get_logger(), "Aborting current goal and accepting new goal");
                    preempted_goal_id_ = goal_handle_->get_goal_id();
                }
            }
        }
        
        RCLCPP_INFO(this->get_logger(), "Goal has been accepted");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse cancel_callback(
        const std::shared_ptr<MoveOneAxisGoalHandle> goal_handle) {
            RCLCPP_INFO(this->get_logger(), "Received cancel request");
            (void)goal_handle;
            return rclcpp_action::CancelResponse::ACCEPT;
        }

    void handle_accepted_callback(const std::shared_ptr<MoveOneAxisGoalHandle> goal_handle) {
        RCLCPP_INFO(this->get_logger(), "Executing the goal");
        execute_goal(goal_handle);
    }

    void execute_goal(const std::shared_ptr<MoveOneAxisGoalHandle> goal_handle) {
        {
            std::lock_guard<std::mutex> lock(mutex_);
            this->goal_handle_ = goal_handle;
        }

        // Get request from goal
        int target_position = goal_handle->get_goal()->target_position;
        int velocity = goal_handle->get_goal()->velocity;

        // Execute the action
        auto result = std::make_shared<MoveOneAxis::Result>();
        auto feedback = std::make_shared<MoveOneAxis::Feedback>();
        rclcpp::Rate loop_rate(1.0);
        int num_move = ceil(abs(target_position-this->current_position_)/velocity);
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
        for (int i = 0; i < num_move; i++) {
            {
                std::lock_guard<std::mutex> lock(mutex_);
                if (goal_handle->get_goal_id() == preempted_goal_id_) {
                    result->final_position = this->current_position_;
                    result->message = "Goal has been preempted";
                    goal_handle->abort(result);
                    return;
                }
            }
            if (goal_handle->is_canceling()) {
                result->final_position = this->current_position_;
                result->message = "Goal has been canceled";
                goal_handle->canceled(result);
                return;
            }
            if (abs(target_position-this->current_position_) < velocity) {
                this->current_position_ += target_position-this->current_position_;
            } else if (target_position > this->current_position_) {
                this->current_position_ += velocity;
            } else {
                this->current_position_ -= velocity;
            }
            feedback->current_position = this->current_position_;
            goal_handle->publish_feedback(feedback);
            RCLCPP_INFO(this->get_logger(), "Current position: %d", this->current_position_);
            loop_rate.sleep();
        }

        // Set final state and return result
        result->final_position = this->current_position_;
        result->message = "Goal has been reached";
        goal_handle->succeed(result);
    }
};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MoveOneAxisServerNode>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();
    // rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}